invalid c code generation with generics, options and type aliases

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->

I couldnt think of a better title.
~~This should probably be a checker error.~~ nope, should work!

**V doctor:**

```
V full version: V 0.4.8 602b097.a3bd5c6
OS: macos, macOS, 15.1.1, 24B2091
Processor: 10 cpus, 64bit, little endian, Apple M4

getwd: /Users/leodev/p/v/fast2
vexe: /Users/leodev/p/v/v/v
vexe mtime: 2024-12-06 07:58:42

vroot: OK, value: /Users/leodev/p/v/v
VMODULES: OK, value: /Users/leodev/.vmodules
VTMP: OK, value: /tmp/v_501

Git version: git version 2.47.0
Git vroot status: weekly.2024.49-29-ga3bd5c64 (4 commit(s) behind V master)
.git/config present: true

CC version: Apple clang version 16.0.0 (clang-1600.0.26.4)
emcc version: N/A
thirdparty/tcc status: thirdparty-macos-arm64 713692d4
```

**What did you do?**
`./v -g -o vdbg cmd/v && ./vdbg repro.v`

```v
import time

struct Foo[T] {
pub:
	value ?T
}

fn test(x ?i64) {
}

fn main() {
	bar := Foo[time.Duration]{
		value: time.Duration(0)
	}

	test(bar.value)
}
```

**What did you expect to see?**

Successful compilation or a checker error

**What did you see instead?**

```
================== C compilation error (from cc): ==============
cc: /tmp/v_501/repro.01JEGF1ABC9SAB8B36WN9B1V2A.tmp.c:7322:13: error: passing '_option_time__Duration' (aka 'struct _option_time__Duration') to parameter of incompatible type '_option_i64' (aka 'struct _option_i64')
cc:  7322 |         main__test(bar.value);
cc:       |                    ^~~~~~~~~
cc: /tmp/v_501/repro.01JEGF1ABC9SAB8B36WN9B1V2A.tmp.c:7316:45: note: passing argument to parameter 'x' here
cc:  7316 | VV_LOCAL_SYMBOL void main__test(_option_i64 x) {
cc:       |                                             ^
cc: 1 error generated.
================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
builder error: 
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```

> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

<sub><a href="https://huly.app/guest/vlang-66f40c4d-a476b54c67-771fdd?token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJsaW5rSWQiOiI2NzU0MzVmMWQyZWY0Y2JjYzQ2Zjc5YzUiLCJndWVzdCI6InRydWUiLCJlbWFpbCI6IiNndWVzdEBoYy5lbmdpbmVlcmluZyIsIndvcmtzcGFjZSI6InctYWxleGFuZGVyLXZsYW5nLTY2ZjQwYzRkLWE0NzZiNTRjNjctNzcxZmRjIn0.wwvN2RxwGhsRNe0wAkQi4UjenDt1lnlkS6XRVE5ZxRY">Huly&reg;: <b>V_0.6-21524</b></a></sub>

--- a/vlib/v/gen/c/assign.v
+++ b/vlib/v/gen/c/assign.v
@@ -54,6 +54,35 @@ fn (mut g Gen) expr_with_opt_or_block(expr ast.Expr, expr_typ ast.Type, var_expr
 	}
 }
 
+// expr_opt_with_alias handles conversion from different option alias type name
+fn (mut g Gen) expr_opt_with_alias(expr ast.Expr, expr_typ ast.Type, ret_typ ast.Type, tmp_var string) {
+	expr_tmp_var := g.new_tmp_var()
+	styp := g.base_type(ret_typ)
+	ret_var := if tmp_var == '' { g.new_tmp_var() } else { tmp_var }
+	line := if tmp_var == '' {
+		s := g.go_before_last_stmt()
+		g.empty_line = true
+		s
+	} else {
+		''
+	}
+	if tmp_var == '' {
+		ret_styp := g.styp(ret_typ).replace('*', '_ptr')
+		g.writeln('${ret_styp} ${ret_var} = {0};')
+	}
+	expr_styp := g.styp(expr_typ).replace('*', '_ptr')
+	g.write('${expr_styp} ${expr_tmp_var} = ')
+	g.expr(expr)
+	g.writeln(';')
+	g.writeln('_option_clone((${option_name}*)&${expr_tmp_var}, (${option_name}*)&${ret_var}, sizeof(${styp}));')
+	if line != '' {
+		g.writeln(line)
+	}
+	if tmp_var == '' {
+		g.write(ret_var)
+	}
+}
+
 // expr_opt_with_cast is used in cast expr when converting compatible option types
 // e.g. ?int(?u8(0))
 fn (mut g Gen) expr_opt_with_cast(expr ast.Expr, expr_typ ast.Type, ret_typ ast.Type) string {
@@ -73,12 +102,7 @@ fn (mut g Gen) expr_opt_with_cast(expr ast.Expr, expr_typ ast.Type, ret_typ ast.
 		g.writeln('${decl_styp} ${past.tmp_var};')
 		is_none := expr is ast.CastExpr && expr.expr is ast.None
 		if expr is ast.CallExpr && expr.return_type.has_flag(.option) {
-			tmp_var := g.new_tmp_var()
-			ret_styp := g.styp(expr.return_type).replace('*', '_ptr')
-			g.write('${ret_styp} ${tmp_var} = ')
-			g.expr(expr)
-			g.writeln(';')
-			g.writeln('_option_clone(&${tmp_var}, (${option_name}*)(&${past.tmp_var}), sizeof(${styp}));')
+			g.expr_opt_with_alias(expr, expr_typ, ret_typ, past.tmp_var)
 		} else {
 			if is_none {
 				g.write('_option_none(&(${styp}[]) {')

--- a/vlib/v/gen/c/assign.v
+++ b/vlib/v/gen/c/assign.v
@@ -55,32 +55,25 @@ fn (mut g Gen) expr_with_opt_or_block(expr ast.Expr, expr_typ ast.Type, var_expr
 }
 
 // expr_opt_with_alias handles conversion from different option alias type name
-fn (mut g Gen) expr_opt_with_alias(expr ast.Expr, expr_typ ast.Type, ret_typ ast.Type, tmp_var string) {
-	expr_tmp_var := g.new_tmp_var()
+fn (mut g Gen) expr_opt_with_alias(expr ast.Expr, expr_typ ast.Type, ret_typ ast.Type, tmp_var string) string {
 	styp := g.base_type(ret_typ)
-	ret_var := if tmp_var == '' { g.new_tmp_var() } else { tmp_var }
-	line := if tmp_var == '' {
-		s := g.go_before_last_stmt()
-		g.empty_line = true
-		s
-	} else {
-		''
-	}
-	if tmp_var == '' {
-		ret_styp := g.styp(ret_typ).replace('*', '_ptr')
-		g.writeln('${ret_styp} ${ret_var} = {0};')
-	}
-	expr_styp := g.styp(expr_typ).replace('*', '_ptr')
-	g.write('${expr_styp} ${expr_tmp_var} = ')
+
+	line := g.go_before_last_stmt().trim_space()
+	g.empty_line = true
+
+	ret_var := g.new_tmp_var()
+	ret_styp := g.styp(ret_typ).replace('*', '_ptr')
+	g.writeln('${ret_styp} ${ret_var} = {0};')
+
+	g.write('_option_clone((${option_name}*)&')
 	g.expr(expr)
-	g.writeln(';')
-	g.writeln('_option_clone((${option_name}*)&${expr_tmp_var}, (${option_name}*)&${ret_var}, sizeof(${styp}));')
-	if line != '' {
-		g.writeln(line)
+	g.writeln(', (${option_name}*)&${ret_var}, sizeof(${styp}));')
+	g.write(line)
+	if g.inside_return {
+		g.write(' ')
 	}
-	if tmp_var == '' {
-		g.write(ret_var)
-	}
+	g.write(ret_var)
+	return ret_var
 }
 
 // expr_opt_with_cast is used in cast expr when converting compatible option types
@@ -93,17 +86,17 @@ fn (mut g Gen) expr_opt_with_cast(expr ast.Expr, expr_typ ast.Type, ret_typ ast.
 	if expr_typ.idx() == ret_typ.idx() && g.table.sym(expr_typ).kind != .alias {
 		return g.expr_with_opt(expr, expr_typ, ret_typ)
 	} else {
-		past := g.past_tmp_var_new()
-		defer {
-			g.past_tmp_var_done(past)
-		}
-		styp := g.base_type(ret_typ)
-		decl_styp := g.styp(ret_typ).replace('*', '_ptr')
-		g.writeln('${decl_styp} ${past.tmp_var};')
-		is_none := expr is ast.CastExpr && expr.expr is ast.None
 		if expr is ast.CallExpr && expr.return_type.has_flag(.option) {
-			g.expr_opt_with_alias(expr, expr_typ, ret_typ, past.tmp_var)
+			return g.expr_opt_with_alias(expr, expr_typ, ret_typ, '')
 		} else {
+			past := g.past_tmp_var_new()
+			defer {
+				g.past_tmp_var_done(past)
+			}
+			styp := g.base_type(ret_typ)
+			decl_styp := g.styp(ret_typ).replace('*', '_ptr')
+			g.writeln('${decl_styp} ${past.tmp_var};')
+			is_none := expr is ast.CastExpr && expr.expr is ast.None
 			if is_none {
 				g.write('_option_none(&(${styp}[]) {')
 			} else {
@@ -120,8 +113,8 @@ fn (mut g Gen) expr_opt_with_cast(expr ast.Expr, expr_typ ast.Type, ret_typ ast.
 				g.inside_opt_or_res = old_inside_opt_or_res
 			}
 			g.writeln(' }, (${option_name}*)(&${past.tmp_var}), sizeof(${styp}));')
+			return past.tmp_var
 		}
-		return past.tmp_var
 	}
 }
 

--- a/vlib/v/gen/c/assign.v
+++ b/vlib/v/gen/c/assign.v
@@ -65,8 +65,18 @@ fn (mut g Gen) expr_opt_with_alias(expr ast.Expr, expr_typ ast.Type, ret_typ ast
 	ret_styp := g.styp(ret_typ).replace('*', '_ptr')
 	g.writeln('${ret_styp} ${ret_var} = {0};')
 
-	g.write('_option_clone((${option_name}*)&')
+	g.write('_option_clone((${option_name}*)')
+	has_addr := expr !in [ast.Ident, ast.SelectorExpr]
+	if has_addr {
+		expr_styp := g.styp(expr_typ).replace('*', '_ptr')
+		g.write('ADDR(${expr_styp}, ')
+	} else {
+		g.write('&')
+	}
 	g.expr(expr)
+	if has_addr {
+		g.write(')')
+	}
 	g.writeln(', (${option_name}*)&${ret_var}, sizeof(${styp}));')
 	g.write(line)
 	if g.inside_return {

--- a/vlib/v/gen/c/fn.v
+++ b/vlib/v/gen/c/fn.v
@@ -2946,7 +2946,11 @@ fn (mut g Gen) ref_or_deref_arg(arg ast.CallArg, expected_type ast.Type, lang as
 		g.write('->val')
 		return
 	} else if expected_type.has_flag(.option) {
-		g.expr_with_opt(arg.expr, arg_typ, expected_type)
+		if arg_sym.info is ast.Alias && expected_type != arg_typ {
+			g.expr_opt_with_alias(arg.expr, arg_typ, expected_type, '')
+		} else {
+			g.expr_with_opt(arg.expr, arg_typ, expected_type)
+		}
 		return
 	} else if arg.expr is ast.ArrayInit {
 		if arg.expr.is_fixed {

--- a/vlib/v/tests/options/option_diff_alias_arg_test.v
+++ b/vlib/v/tests/options/option_diff_alias_arg_test.v
@@ -1,0 +1,18 @@
+import time
+
+struct Foo[T] {
+pub:
+	value ?T
+}
+
+fn t(x ?i64) i64 {
+	return x or { -1 }
+}
+
+fn test_main() {
+	bar := Foo[time.Duration]{
+		value: time.Duration(0)
+	}
+
+	assert t(bar.value) == 0
+}

